### DISCUSSION TITLE
Using LeanDojo in Google Colab

### DISCUSSION THREAD

**Discussion by ZIYU-DEEP** [lean-dojo/LeanDojo#154] [Q&A] [answered]
[![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1n8fMBM_jtQKQ673Q13AJoGZhEJZA4EIe?usp=sharing)

I am attempting to use the package in colab notebook but facing some minor issues.

Setup steps:
```python
## Preparation
import os

os.environ['CONTAINER'] = 'native'
os.environ['VERBOSE'] = '1'
os.environ['GITHUB_ACCESS_TOKEN'] = input()

## elan and lean installation
# Download the binary
!wget https://github.com/leanprover/elan/releases/download/v3.1.1/elan-x86_64-unknown-linux-gnu.tar.gz

# Extract the content
!tar -xzf elan-x86_64-unknown-linux-gnu.tar.gz

# Move the the local bin directory
!mv elan-init /usr/local/bin/elan

# Verify the downloaded elan
!/usr/local/bin/elan --version

# Download Lean
!/usr/local/bin/elan default stable

# Check the lean version
!elan which lean

# Add the above path to the environment
lean_bin_path = '/root/.elan/toolchains/stable/bin'
os.environ['PATH'] += ':' + lean_bin_path

# Check the lean version
!lean --version

## Test LeanDojo
!pip install lean-dojo
```

Initial issue: running as the root
```
from lean_dojo import *
```
This will trigger an error due to `__init__.py` raising an exception for running as root. It seems to be a bit cumbersome to run as a non-root user in colab, so i just comment out the error-raising lines in order to proceed. What would be some better practices to handle this issue?

Subsequent issue: tracing repository stalls
```
repo = LeanGitRepo(
    'https://github.com/yangky11/lean4-example', 
    '3f8c5eb303a225cdef609498b8d87262e5ef344b'  # Using the Lean version 4.7.0
    )

trace(repo, dst_dir='traced_lean4-example')
```
Below are the logs:
```
2024-04-10 23:23:06.240 | INFO     | lean_dojo.data_extraction.trace:get_traced_repo_path:79 - Tracing LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2024-04-10 23:23:06.243 | DEBUG    | lean_dojo.data_extraction.trace:get_traced_repo_path:81 - Working in the temporary directory /tmp/tmpak727a5s
2024-04-10 23:23:06.246 | DEBUG    | lean_dojo.data_extraction.lean:clone_and_checkout:475 - Cloning LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2024-04-10 23:23:06.626 | DEBUG    | lean_dojo.data_extraction.trace:_trace:33 - Tracing LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2024-04-10 23:23:06.637 | DEBUG    | lean_dojo.container:run:184 - NUM_PROCS=2 python build_lean4_repo.py lean4-example
```
And it looks like the process gets stuck at `NUM_PROCS=2 python build_lean4_repo.py lean4-example`. I am wondering what are some suggestions to help diagnose the issue. Many thanks.

**Comment by yangky11** [comment]
> so i just comment out the error-raising lines in order to proceed.

I haven't used Colab a lot before. How to edit a source file on Colab?

**Reply by ZIYU-DEEP** [reply]
Hi Kaiyu – directly clicking the filename (`/usr/local/lib/python3.10/dist-packages/lean_dojo/__init__.py`) in the error message should open it as a editable file in the right side pane; or using the `files` tab on the left sidebar to navigate to the file.

**Comment by yangky11** [comment]
It looks like Colab cannot find `elan`, `lake`, etc. I got an error when running
```python
os.system("lake --help")
```

@ZIYU-DEEP Do you have the same issue?

**Reply by ZIYU-DEEP** [reply]
Hi Kaiyu, it looks to be working in my case.

```python
os.system('lake --help')
```
```
>>> 0
```

To check the output:
```python
import subprocess
result = subprocess.run(['lake', '--help'], stdout=subprocess.PIPE, text=True)
print(result.stdout)
```
or:
```
!lake --help
```
Both can give:

```
>>> Lake version 5.0.0-6fce8f7 (Lean version 4.7.0)
>>>
>>> USAGE:
>>>   lake [OPTIONS] <COMMAND>
>>>
>>> COMMANDS:
>>>   new <name> <temp>     create a Lean package in a new directory
>>>   init <name> <temp>    create a Lean package in the current directory
>>>   build <targets>...    build targets
>>>   exe <exe> <args>...   build an exe and run it in Lake's environment
>>>   clean                 remove build outputs
>>>   env <cmd> <args>...   execute a command in Lake's environment
>>>   update                update dependencies and save them to the manifest
>>>   upload <tag>          upload build artifacts to a GitHub release
>>>   script                manage and run workspace scripts
>>>   scripts               shorthand for `lake script list`
>>>   run <script>          shorthand for `lake script run`
>>>   serve                 start the Lean language server
>>>
>>> OPTIONS:
>>>   --version             print version and exit
>>>   --help, -h            print help of the program or a command and exit
>>>   --dir, -d=file        use the package configuration in a specific directory
>>>   --file, -f=file       use a specific file for the package configuration
>>>   --quiet, -q           hide progress messages
>>>   --verbose, -v         show verbose information (command invocations)
>>>   --lean=cmd            specify the `lean` command used by Lake
>>>   -K key[=value]        set the configuration file option named key
>>>   --old                 only rebuild modified modules (ignore transitive deps)
>>>   --rehash, -H          hash all files for traces (do not trust `.hash` files)
>>>   --update, -U          update manifest before building
>>>   --reconfigure, -R     elaborate configuration files instead of using OLeans
>>>
>>> See `lake help <command>` for more information on a specific command.
```

@yangky11 Here is the colab link: [![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1jiS_ATivXz3NpvFoUrHc8p36CGYHpti7?usp=sharing).

**Comment by ZIYU-DEEP** [comment, marked as answer]
[![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1jiS_ATivXz3NpvFoUrHc8p36CGYHpti7?usp=sharing)

@yangky11 the environment works! here is a [colab](https://colab.research.google.com/drive/1jiS_ATivXz3NpvFoUrHc8p36CGYHpti7?usp=sharing) version of [demo-lean4.ipynb](https://github.com/lean-dojo/LeanDojo/blob/405513a55181bb0552e41833c13862f3a03c9139/scripts/demo-lean4.ipynb).

**Reply by AmitAminov** [reply]
I am getting the following exception when running the demo in Google Colab.
The error accurs when running the line: traced_repo = trace(repo, dst_dir='traced_lean4-example')

2025-07-19 09:42:10.993 | INFO     | lean_dojo.data_extraction.trace:get_traced_repo_path:79 - Tracing LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2025-07-19 09:42:10.998 | DEBUG    | lean_dojo.data_extraction.trace:get_traced_repo_path:81 - Working in the temporary directory /content/drive/MyDrive/tmp/tmp1ox_5w89
2025-07-19 09:42:11.467 | DEBUG    | lean_dojo.data_extraction.lean:clone_and_checkout:475 - Cloning LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2025-07-19 09:42:13.046 | DEBUG    | lean_dojo.data_extraction.trace:_trace:33 - Tracing LeanGitRepo(url='https://github.com/yangky11/lean4-example', commit='3f8c5eb303a225cdef609498b8d87262e5ef344b')
2025-07-19 09:42:13.521 | DEBUG    | lean_dojo.container:run:184 - NUM_PROCS=8 python build_lean4_repo.py lean4-example
---------------------------------------------------------------------------
CalledProcessError                        Traceback (most recent call last)
[/tmp/ipython-input-16-635697842.py](https://localhost:8080/#) in <cell line: 0>()
      1 # Trace the repo; this may take approximately 1.2 hrs
----> 2 traced_repo = trace(repo, dst_dir='traced_lean4-example')

7 frames
[/usr/local/lib/python3.11/dist-packages/lean_dojo/data_extraction/trace.py](https://localhost:8080/#) in trace(repo, dst_dir, build_deps)
    113         ), f"The destination directory {dst_dir} already exists."
    114 
--> 115     cached_path = get_traced_repo_path(repo, build_deps)
    116     logger.info(f"Loading the traced repo from {cached_path}")
    117     traced_repo = TracedRepo.load_from_disk(cached_path, build_deps)

[/usr/local/lib/python3.11/dist-packages/lean_dojo/data_extraction/trace.py](https://localhost:8080/#) in get_traced_repo_path(repo, build_deps)
     80         with working_directory() as tmp_dir:
     81             logger.debug(f"Working in the temporary directory {tmp_dir}")
---> 82             _trace(repo, build_deps)
     83             traced_repo = TracedRepo.from_traced_files(tmp_dir / repo.name, build_deps)
     84             traced_repo.save_to_disk()

[/usr/local/lib/python3.11/dist-packages/lean_dojo/data_extraction/trace.py](https://localhost:8080/#) in _trace(repo, build_deps)
     55                 "Failed to build Lean 4 without Docker. See https://leandojo.readthedocs.io/en/latest/user-guide.html#advanced-running-within-docker."
     56             )
---> 57         raise ex
     58 
     59 

[/usr/local/lib/python3.11/dist-packages/lean_dojo/data_extraction/trace.py](https://localhost:8080/#) in _trace(repo, build_deps)
     43 
     44     try:
---> 45         container.run(
     46             cmd,
     47             create_mounts(mts),

[/usr/local/lib/python3.11/dist-packages/lean_dojo/container.py](https://localhost:8080/#) in run(self, command, mounts, envs, as_current_user, capture_output, cpu_limit, memory_limit, work_dir)
    192 
    193         with working_directory(work_dir):
--> 194             execute(cmd, capture_output=capture_output)
    195 
    196         self._unmount_files(mounts)

[/usr/local/lib/python3.11/dist-packages/lean_dojo/utils.py](https://localhost:8080/#) in execute(cmd, capture_output)
    114             logger.info(ex.stdout.decode())
    115             logger.error(ex.stderr.decode())
--> 116         raise ex
    117     if not capture_output:
    118         return None

[/usr/local/lib/python3.11/dist-packages/lean_dojo/utils.py](https://localhost:8080/#) in execute(cmd, capture_output)
    109     """
    110     try:
--> 111         res = subprocess.run(cmd, shell=True, capture_output=capture_output, check=True)
    112     except subprocess.CalledProcessError as ex:
    113         if capture_output:

[/usr/lib/python3.11/subprocess.py](https://localhost:8080/#) in run(input, capture_output, timeout, check, *popenargs, **kwargs)
    569         retcode = process.poll()
    570         if check and retcode:
--> 571             raise CalledProcessError(retcode, process.args,
    572                                      output=stdout, stderr=stderr)
    573     return CompletedProcess(process.args, retcode, stdout, stderr)

CalledProcessError: Command 'NUM_PROCS=8 python build_lean4_repo.py lean4-example' returned non-zero exit status 1.
